tsdb/index/tsi1/index.go

```diff
@@ -135,1277 +146,770 @@ func (i *Index) Open() error {
-func (i *Index) Manifest() *Manifest {
-	m := &Manifest{
-		Levels:  i.levels,
-		Files:   make([]string, len(i.fileSet.files)),
-		Version: i.version,
-	}
-
-	for j, f := range i.fileSet.files {
-		m.Files[j] = filepath.Base(f.Path())
-	}
-
-	return m
+// partition returns the appropriate Partition for a provided series key.
+func (i *Index) partition(key []byte) *Partition {
+	return i.partitions[int(xxhash.Sum64(key)&(i.PartitionN-1))]
 }
 
-// writeManifestFile writes the manifest to the appropriate file path.
-func (i *Index) writeManifestFile() error {
-	return WriteManifestFile(i.ManifestPath(), i.Manifest())
+// partitionIdx returns the index of the partition that key belongs in.
+func (i *Index) partitionIdx(key []byte) int {
+	return int(xxhash.Sum64(key) & (i.PartitionN - 1))
 }
 
-// WithLogger sets the logger for the index.
-func (i *Index) WithLogger(logger *zap.Logger) {
-	i.logger = logger.With(zap.String("index", "tsi"))
+// availableThreads returns the minimum of GOMAXPROCS and the number of
+// partitions in the Index.
+func (i *Index) availableThreads() int {
+	n := runtime.GOMAXPROCS(0)
+	if len(i.partitions) < n {
+		return len(i.partitions)
+	}
+	return n
 }
 
 // SetFieldSet sets a shared field set from the engine.
 func (i *Index) SetFieldSet(fs *tsdb.MeasurementFieldSet) {
-	i.mu.Lock()
-	i.fieldset = fs
-	i.mu.Unlock()
-}
-
-// RetainFileSet returns the current fileset and adds a reference count.
-func (i *Index) RetainFileSet() *FileSet {
-	i.mu.RLock()
-	fs := i.retainFileSet()
-	i.mu.RUnlock()
-	return fs
-}
-
-func (i *Index) retainFileSet() *FileSet {
-	fs := i.fileSet
-	fs.Retain()
-	return fs
-}
-
-// FileN returns the active files in the file set.
-func (i *Index) FileN() int { return len(i.fileSet.files) }
-
-// prependActiveLogFile adds a new log file so that the current log file can be compacted.
-func (i *Index) prependActiveLogFile() error {
-	// Open file and insert it into the first position.
-	f, err := i.openLogFile(filepath.Join(i.Path, FormatLogFileName(i.nextSequence())))
-	if err != nil {
-		return err
+	for _, p := range i.partitions {
+		p.SetFieldSet(fs)
 	}
-	i.activeLogFile = f
-
-	// Prepend and generate new fileset.
-	i.fileSet = i.fileSet.PrependLogFile(f)
+}
 
-	// Write new manifest.
-	if err := i.writeManifestFile(); err != nil {
-		// TODO: Close index if write fails.
-		return err
+// FieldSet returns the assigned fieldset.
+func (i *Index) FieldSet() *tsdb.MeasurementFieldSet {
+	if len(i.partitions) == 0 {
+		return nil
 	}
-
-	return nil
+	return i.partitions[0].FieldSet()
 }
 
-// ForEachMeasurementName iterates over all measurement names in the index.
+// ForEachMeasurementName iterates over all measurement names in the index,
+// applying fn. Note, the provided function may be called concurrently, and it
+// must be safe to do so.
+//
+// It returns the first error encountered, if any.
 func (i *Index) ForEachMeasurementName(fn func(name []byte) error) error {
```

Comment:
Looks like we've will lose the sorting of measurement names now.  Are they still iterated in sorted order?

Reply:
Yes, partitions changed the sorting. It will also cause duplicates. I need to swap this out for a `MeasurementIterator` which will prevent both issues.